storage/innobase/handler/ha_innodb.cc

```diff
@@ -18360,6 +18360,16 @@ innobase_fts_find_ranking(FT_INFO* fts_hdl, uchar*, uint)
 	return(fts_retrieve_ranking(result, ft_prebuilt->fts_doc_id));
 }
 
+/** Update a field that is protected by buf_pool.mutex */
+template<typename T>
+static void innodb_buf_pool_update(THD *thd, st_mysql_sys_var *,
+                                   void *val, const void *save)
+{
+  mysql_mutex_lock(&buf_pool.mutex);
+  *static_cast<T*>(val)= *static_cast<const T*>(save);
+  mysql_mutex_unlock(&buf_pool.mutex);
+}
```

Comment:
Looks correct. Some things to note here. I would not recommend any change right now but let's just note in case any user comes back with similar observation.

1. We could possibly get blocked for some time here waiting for buffer pool mutex. User might observe the difference here as these modifications are generally considered asynchronous.

2. A related point is that we are also likely holding some configuration mutex from server here and if we block here it might block update to other configuration variables. I don't see a possibility of deadlock unless we ask back for configuration information while holding the buffer pool mutex.

Reply:
Right, we would be holding `LOCK_global_system_variables` while waiting for the `buf_pool.mutex`. In some places, where the wait is expected to be much longer, we are releasing `LOCK_global_system_variables` while waiting. This could open an opportunity of further race conditions. Here I do not see the need for it; `buf_pool.mutex` should never be held for an extensive period of time. The same can’t currently be said about `dict_sys.latch`, for instance.